### ISSUE TITLE
CTRL+D delete string shortcut removed in jupyter 4.0.0

### ISSUE THREAD

**Issue by mahashiah** [jupyterlab/jupyterlab#14548]
Deleting a line (string) has become an impossible task on my keyboard since developers removed the Ctrl+D shortcut in version 4.0. I must commend the developers for making it exceptionally challenging to utilize Jupyter on a 60% mechanical keyboard lacking a dedicated delete key. It seems to be an erroneous assumption that every keyboard possesses such a key, which is evidently not the case. Moreover, you have imposed the compulsory assignment of delete as a vital key while assigning Ctrl+D to a new feature that is essentially futile. It would have been far more practical to allocate Ctrl+F for word searches, as you did for Ctrl+D. How can I delete a row in Jupyter 4.0 when the Ctrl+D shortcut is invariably linked to an unhelpful search function? To be frank, I had a favorable opinion of Jupyter, but now I find it necessary to downgrade and revert to a previous version, abandoning 4.0 entirely.

**Comment by JasonWeill** [comment]
@mahashiah Thank you for your suggestion! We apologize for the regression and we welcome contributions to reinstate it.

**Comment by krassowski** [comment]
A few clarifications to dissect the issue:
- there was no search keybinding for <kbd>Ctrl</kbd> + <kbd>D</kbd> added in by JupyterLab team in JupyterLab 4.0. The search widget keybindings are defined here:
       https://github.com/jupyterlab/jupyterlab/blob/94a27fb987afddaf75d65265d825f24774247598/packages/documentsearch-extension/schema/plugin.json#L35-L56
- the change is a result of migration from CodeMirror5 to CodeMirror6 which changed the default keybindings. The new keybinding is <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>K</kbd> (defined in [`codemirror/commands` package, here](https://github.com/codemirror/commands/blob/f17b5b889801959cdc01435ba2b5f23c68ecc841/src/commands.ts#L895)) but it conflicts with the JupyterLab shortcut to open Table of Contents sidebar
  - CodeMirror does define [`Mod-d` for select next occurrence]( https://github.com/codemirror/search/blob/3c1346e213a9635f56f92b23066ea0638319de25/src/search.ts#L583) command (which we do not need in most editors as described later)

How we can address this?
- the default keymap of CodeMirror in JupyterLab can be amended to reduce friction for users migrating from 3.x; this likely should happen either;
   - in [`CodeMirrorEditorFactory`](https://github.com/jupyterlab/jupyterlab/blob/94a27fb987afddaf75d65265d825f24774247598/packages/codemirror/src/factory.ts#L21) if we want to hard-code the shortcut, or
   - [`getDefaultExtensions`](https://github.com/jupyterlab/jupyterlab/blob/94a27fb987afddaf75d65265d825f24774247598/packages/codemirror/src/extension.ts#LL593C19-L593C39) if we want to expose this as customisable option
- the search keybindings keymap should probably be removed from inline editors (those are added [here](https://github.com/jupyterlab/jupyterlab/blob/94a27fb987afddaf75d65265d825f24774247598/packages/codemirror/src/factory.ts#L48)) because JupyterLab has a custom search & replace implementation - now independent of CodeMirror; the downstream users will still be able to enable the search by passing `extensions` in options (which we may need to do in JSON setting editor as this one does not implement `documentsearch` provider)

Finally, @mahashiah:
- please be reminded that Project Jupyter contributors (including contributions to our GitHub discussions) are expected to follow the [Code of Conduct](https://github.com/jupyter/governance/blob/main/conduct/code_of_conduct.md), in particular I would like to highlight the section about [expected behaviour](https://github.com/jupyter/governance/blob/main/conduct/code_of_conduct.md#expected-behavior).
- in future please feel welcome to contribute by trying out beta/release candidate versions - critical feedback is easier to address at these stages, and these can usually be trialled without installation on Binder.

**Comment by mahashiah** [comment]
@krassowski  krassowski thank you for your dedicated time and clarifying that I did not intend to show a sarcastic or provocative tone at any time, it was translation by gpt. what happened was a critical error, 60% of the keyboards don't have the delete key, just backspace, jupyterlab 3.6 has the default shortcut to delete in crtl+D, in version 4.0.0 released this shortcut was changed, when I installed the version 4.0 in my environment crtl+D was assigned a word search function, when pressing crtl+D it selected a word and found all the same words (as if it were crtl+F) I accessed the settings part to change the shortcuts and this search The function had crtl+D frozen, making it impossible to assign crtl+D to any other shortcut, including delete. I can redo the process with recording and if the same problem occurs, bring the video here. when closed we can, if possible, delete the topic.

**Comment by SysuJayce** [comment]
So can we have ctrl+D to delete line again ?

**Comment by andrii-i** [comment]
Thank you for interest in this issue @SysuJayce. It seems that people participating in the discussion agree that the default keymap of CodeMirror in JupyterLab can be amended and this issue is planned to be resolved within current 4.0.x milestone. Contributions to resolve it are also welcomed.

**Comment by SysuJayce** [comment]
> Thank you for interest in this issue @SysuJayce. It seems that people participating in the discussion agree that the default keymap of CodeMirror in JupyterLab can be amended and this issue is planned to be resolved within current 4.0.x milestone. Contributions to resolve it are also welcomed.

Thank you @andrii-i . I also migrate from 3.x and I like the keyboard shortcut of 3.x. It is great to know that we can change the default keymap in 4.x 👍

**Comment by RafalSkolasinski** [comment]
I would really love if it would be possible to have both shortcuts for multi select and full line removal. Preferably configurable to have mutli select on ctrl+d and line removal on ctrl+shift+k to match VS Code defaults

**Comment by kelvinAI** [comment]
For people who uses normal keyboards which have the Delete key, `Shift+Delete` seems to do the job of deleting the line the cursor is on, at least on Jupyter 4.0.3 with default settings. However I'd still advocate for `Ctrl+D` as it is consistent with popular editors such as VS Code etc and it's easier to execute.

**Comment by y-he2** [comment]
> For people who uses normal keyboards which have the Delete key, `Shift+Delete` seems to do the job of deleting the line the cursor is on, at least on Jupyter Lab 4.0.3 with default settings. That being said I'd still advocate for `Ctrl+D` as it is consistent with popular editors such as VS Code etc and it's easier to execute.

It seems Shift+Del is equivalent as Line Cut in standard PC:
https://en.wikipedia.org/wiki/Cut,_copy,_and_paste#Common_keyboard_shortcuts
so a Ctrl-X would prob do the same trick on Windows. Although beaware it could mess up your current clip-board, which personally is the biggest reason to use a separate hotkey (eg. Ctrl-D) for "line delete".

Also I would like to contribute to this issue. (Ctrl-D is like a religious thing to me when rapid prototyping.) Could someone point a direction where to look into, or is there any meeting I could join regarding this?

**Comment by krassowski** [comment]
Hi @y-he2 are you still interested in contributing? Let me know if you need further pointers.

**Comment by y-he2** [comment]
> Hi @y-he2 are you still interested in contributing? Let me know if you need further pointers.

Absolutely, Im pretty determined to solve this and https://github.com/jupyterlab/jupyterlab/issues/4778.
(Two draw back may be I have never worked in the framework used by Jupyter 4 before, also I have limited time.)

It would be easiest if I can join some meeting or if we can do some "Discord" to get me started.

**Comment by JasonWeill** [comment]
@y-he2 Thank you for your interest! We don't have an official Discord, but you can join our Gitter chat server here: https://app.gitter.im/#/room/#jupyter_jupyter:gitter.im

**Comment by krassowski** [comment]
We also have a [contributing guide](https://jupyterlab.readthedocs.io/en/stable/developer/contributing.html) - don't feel obliged to read it all, but it will likely have answers to most of questions about setup and contributing, basically if you can't find something in there after 15 minutes shout out in gitter.

**Comment by y-he2** [comment]
Right, thanks for the info, will join and have a look at the guide.

**Comment by y-he2** [comment]
Right, thanks for the info, will join and have a look at the guide.

**Comment by SysuJayce** [comment]
any progress?

**Comment by y-he2** [comment]
I have setup the environment and started to study the code. Have been busy all week days as usual.
Trying to figure out next how to do:

> "or
 [getDefaultExtensions] https://github.com/jupyterlab/jupyterlab/blob/94a27fb987afddaf75d65265d825f24774247598/packages/codemirror/src/extension.ts#LL593C19-L593C39) if we want to expose this as customisable option"

and how these factories works with custom hotkey stuff. 

Can keep us updated with further detail when I have more.

**Comment by jtrauer** [comment]
Desperately need ctrl-D back.

**Comment by dimi-tree** [comment]
Is this issue being worked on? I can help if no one is looking at it.

**Comment by marius311** [comment]
I know this sounds dramatic but as someone that was rght on the edge of dropping JupyterLab entirely in favor VSCode's Jupyter because of its superior editor functionality, this ended up being a straw the broke the camels back and I've switched. Just my personal feedback if its helpful for devs to prioritze work. I could see myself swtiching back bc other stuff is better here, but I currently feel handcuffed in the editor (worse multicursor, less productivity plugins, ofc the now-missing Ctrl+D, etc...)

**Comment by JasonWeill** [comment]
@y-he2 @marius311 Thank you for your continuing engagement with this issue! You can join us in our Matrix (Gitter, Element) chat room about JupyterLab here: https://matrix.to/#/#jupyterlab_jupyterlab:gitter.im

Our developers would be happy to review a pull request to help fix this issue.

**Comment by y-he2** [comment]
> I know this sounds dramatic but as someone that was rght on the edge of dropping JupyterLab entirely in favor VSCode's Jupyter because of its superior editor functionality, this ended up being a straw the broke the camels back and I've switched. Just my personal feedback if its helpful for devs to prioritze work. I could see myself swtiching back bc other stuff is better here, but I currently feel handcuffed in the JupyterLab editor (worse multicursor, less productivity plugins, ofc the now-missing Ctrl+D, etc...)

I think we all understand the importance of this issue, just it may not be as easy to fix as it seems.

**Comment by krassowski** [comment]
> Our developers

To clarify, there is a group of folks who maintain and/or contribute to JupyterLab project, some of whom are volunteers and in general there is no ownership of developers; Project Jupyter itself (as far as I know) does not employ any developers (though maybe I am wrong? this is probably a topic for a different discussion). That said, it is true that (part of) JupyterLab team expressed intent to prioiritize review of pull requests fixing regressions in 4.0 over other pull requests.

As previously stated, I am happy to guide any contributor in my free time.

> just it may not be as easy to fix as it seems

II think it would be fair to say that it is quite time consuming  to fix as it involves both codemirror and JupyterLab shortcut systems, and there is so much time one can sacrifice from the family time in the evenings :man_shrugging:

**Comment by JasonWeill** [comment]
I agree with @krassowski that JupyterLab is a community-run project. We have a large community of contributors, and we welcome contributions from anyone who can help improve our software. The best way to have a positive impact on the community is to work on and contribute pull requests. Adding comments and votes on issues makes them more visible, but there is no core dev team that uses them to prioritize work.

**Comment by y-he2** [comment]
> @y-he2 @marius311 Thank you for your continuing engagement with this issue! You can join us in our Matrix (Gitter, Element) chat room about JupyterLab here: https://matrix.to/#/#jupyterlab_jupyterlab:gitter.im
> 
> Our developers would be happy to review a pull request to help fix this issue.

Thanks for the links again, I was trying to pointing @dimi-tree to it but didnt pasted it.

More over I was thinking maybe creating a thread in that Gitter chat room, or a Discord server _specifically for this issue and #4778_ so we wont spam the main chat, before the pull request. Because I feel like we would need to gather some work force for these two issues, and it would be easier to communicate by just join the specific server for it for discussions. Since it may be a bit too much for solos alone.
This was the essential point I pinged you guys, because I didnt knew if it would mess things up by creating separate communication channels. @krassowski @JasonWeill 

> it is quite time consuming to fix as it involves both codemirror and JupyterLab shortcut systems, and there is so much time one can sacrifice from the family time in the evenings
Yes I totally agree and that was part of the message I was trying to forward, apologies for any misunderstanding that may be caused by any miscommunication.
